### ISSUE TITLE
Dialog won't unmount when dismissed while mobile keyboard is open

### ISSUE THREAD

**Issue by ericwang401** [tailwindlabs/headlessui#2032]
**What package within Headless UI are you using?**

@headlessui/react

**What version of that package are you using?**

v1.7.4

**What browser are you using?**

Android Brave Web Browser
Application version: Brave 1.45.123, Chromium 107.0.5304.110
Operating System: Android 10; Build/QP1A. 190711.020

**Reproduction URL**

[Codesandbox](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-bjs4vr?file=%2Fsrc%2FApp.tsx&selection=%5B%7B%22endColumn%22%3A64%2C%22endLineNumber%22%3A11%2C%22startColumn%22%3A7%2C%22startLineNumber%22%3A11%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522cl9zkf3e9000alqiphi669wk8%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%252C%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%252C%2522%252Fsrc%252FApp.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522clan6rf4001ke376i80a287gn%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clan6re0c01g3376ioybsxh1c%2522%252C%2522isMinimized%2522%253Atrue%257D%255D%257D)

**Describe your issue**

I'm trying to make a dialog using the Headless UI Dialog component and the Transition component so that the dialog looks like this on desktop:
![image](https://user-images.githubusercontent.com/37554696/202823962-e73949e2-4ee2-4ae0-aa25-5a2d6141510b.png)

And like this on mobile:
![image](https://user-images.githubusercontent.com/37554696/202823989-afa25c16-ee48-4f08-9456-b47851502bd0.png)

However, I found out that the dialog component doesn't unmount in this specific scenario that is **VERY** easy to trigger. This prevents the user from being able to interact with anything as the dialog component exists in the DOM but isn't visible in the viewport.

Reproduction instructions:
1. User is on **mobile** (VERY IMPORTANT). Use a mobile web browser, the devtools in desktop chromium doesn't simulate this bug
2. User opens the dialog
3. User focuses the textbox and the mobile keyboard is visible
4. User then clicks the backdrop behind the dialog, dismissing the dialog. The user is performing this while the mobile keyboard is still active
5. Dialog dismisses and transitions out of view. Though the component doesn't unmount and blocks any interaction on any components underneath (buttons, text inputs, etc).

Video demonstration:
https://youtube.com/shorts/Qhr1_azemaM?feature=share

The actual result I want is when the user dismisses while the virtual keyboard is open, the dialog should transition out of view and unmount so that underlying elements can be interacted with.

I have already asked about this problem, [and there was a workaround (though really nasty workaround) suggested by a answerer on Stackoverflow](https://stackoverflow.com/questions/74497444/dialog-wont-unmount-when-dismissed-while-mobile-keyboard-is-open/74497766#74497766). This workaround doesn't always work with 0 milliseconds. I have to set it to around 100+ milliseconds for the component to properly unmount in this very specific situation.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your bug report!
Much appreciated! 🙏

The Transition component has to transition out completely, once that is done the Modal will unmount. Fwiw, I cannot reproduce this on iOS.

Now, for the actual issue, we are using the native `transitionrun`, `transitionend` and `transitioncancel` events. I know they existed for a long time, but some of them were only implemented recently (some people reported issues before on iOS 12.x).

Can you reproduce the issue if you remove the transitions (and Transition component) altogether? [CodeSandbox](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-forked-oo18kq?file=%2Fsrc%2Fcomponents%2Felements%2FModal.tsx&selection=%5B%7B%22endColumn%22%3A15%2C%22endLineNumber%22%3A18%2C%22startColumn%22%3A15%2C%22startLineNumber%22%3A18%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522clan5wgyn00pil1ipfe9p5ja0%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522claqt6rae00em3b5y4pnhh5pd%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D) — Preview: https://oo18kq-5173.preview.csb.app/

**Comment by ericwang401** [comment]
> Hey! Thank you for your bug report! Much appreciated! 🙏
> 
> The Transition component has to transition out completely, once that is done the Modal will unmount. Fwiw, I cannot reproduce this on iOS.
> 
> Now, for the actual issue, we are using the native `transitionrun`, `transitionend` and `transitioncancel` events. I know they existed for a long time, but some of them were only implemented recently (some people reported issues before on iOS 12.x).
> 
> Can you reproduce the issue if you remove the transitions (and Transition component) altogether? [CodeSandbox](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-forked-oo18kq?file=%2Fsrc%2Fcomponents%2Felements%2FModal.tsx&selection=%5B%7B%22endColumn%22%3A15%2C%22endLineNumber%22%3A18%2C%22startColumn%22%3A15%2C%22startLineNumber%22%3A18%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522clan5wgyn00pil1ipfe9p5ja0%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522claqt6rae00em3b5y4pnhh5pd%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D) — Preview: https://oo18kq-5173.preview.csb.app/

Hello. Thanks for the reply!

**== this part below is a little off topic but I hope there is some insight to the main bug from this experience==**
I kind of see how transition works. I'm curious how the textbox being focused causes this bug. Before this main bug, I had another "bug" with the transitions when the modal pops into view on mobile screens. The animation would stutter and not animate properly from below to in the viewport on mobile but immediately pop in--as if there wasn't a transition--because the textbox in the dialog component was automatically focused by the [FocusTrap component](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/dialog/dialog.tsx#L345). As a workaround to fix the stuttering transitions, I added a hidden input element (in line 57 in `Modal.tsx` in the original Codesandbox) and set the `initialFocus` to the ref of the hidden input. And this "tricked" the `<Dialog>` component to not autoFocus and make the animation stutter. This other bug should be reproducible on Chrome desktop if you resize the width of your screen to <640px or the size of a phone. **In addition, the main bug discussed in this issue is still present without the use of the hidden input element.**

In the video, to prevent the (main) unmounting bug from happening, after I have focused the textbox, I have to unfocus it by clicking somewhere inside the modal. Then, I have to wait a little bit before clicking the backdrop to dismiss so that the modal can properly unmount.

[Video Demonstration of the stuttering animation](https://youtube.com/shorts/MMcLGFhW5nw?feature=share)

[Stuttering Dialog Codesandbox (this is without the hidden input element)](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-forked-x7j9sd?file=%2Fsrc%2Fcomponents%2Felements%2FModal.tsx&selection=%5B%7B%22endColumn%22%3A40%2C%22endLineNumber%22%3A56%2C%22startColumn%22%3A40%2C%22startLineNumber%22%3A56%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522clan5wgyn00pil1ipfe9p5ja0%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522clasrdalq006o3b6iowtn6unl%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D)


**==Back to the main bug in this issue==**
I tried your Codesandbox link (https://oo18kq-5173.preview.csb.app/) and the unmounting bug isn't present this time!

So seems like this is an issue with the `<Transition>` component? I'll maybe have to also do some research on the implementation in HeadlessUI.

Here's the [video demonstration](https://youtube.com/shorts/I9xu3uHNbpQ?feature=share) with your Codesandbox link that omitted the transitions.


This is a very peculiar problem, and I appreciate your help on this! Thanks again!

**Comment by RobSchilderr** [comment]
> video demonstration

Let me know if you solved this!

**Comment by RobSchilderr** [comment]
> > Hey! Thank you for your bug report! Much appreciated! 🙏
> > The Transition component has to transition out completely, once that is done the Modal will unmount. Fwiw, I cannot reproduce this on iOS.
> > Now, for the actual issue, we are using the native `transitionrun`, `transitionend` and `transitioncancel` events. I know they existed for a long time, but some of them were only implemented recently (some people reported issues before on iOS 12.x).
> > Can you reproduce the issue if you remove the transitions (and Transition component) altogether? [CodeSandbox](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-forked-oo18kq?file=%2Fsrc%2Fcomponents%2Felements%2FModal.tsx&selection=%5B%7B%22endColumn%22%3A15%2C%22endLineNumber%22%3A18%2C%22startColumn%22%3A15%2C%22startLineNumber%22%3A18%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522clan5wgyn00pil1ipfe9p5ja0%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522claqt6rae00em3b5y4pnhh5pd%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D) — Preview: https://oo18kq-5173.preview.csb.app/
> 
> Hello. Thanks for the reply!
> 
> **== this part below is a little off topic but I hope there is some insight to the main bug from this experience==** I kind of see how transition works. I'm curious how the textbox being focused causes this bug. Before this main bug, I had another "bug" with the transitions when the modal pops into view on mobile screens. The animation would stutter and not animate properly from below to in the viewport on mobile but immediately pop in--as if there wasn't a transition--because the textbox in the dialog component was automatically focused by the [FocusTrap component](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/dialog/dialog.tsx#L345). As a workaround to fix the stuttering transitions, I added a hidden input element (in line 57 in `Modal.tsx` in the original Codesandbox) and set the `initialFocus` to the ref of the hidden input. And this "tricked" the `<Dialog>` component to not autoFocus and make the animation stutter. This other bug should be reproducible on Chrome desktop if you resize the width of your screen to <640px or the size of a phone. **In addition, the main bug discussed in this issue is still present without the use of the hidden input element.**
> 
> In the video, to prevent the (main) unmounting bug from happening, after I have focused the textbox, I have to unfocus it by clicking somewhere inside the modal. Then, I have to wait a little bit before clicking the backdrop to dismiss so that the modal can properly unmount.
> 
> [Video Demonstration of the stuttering animation](https://youtube.com/shorts/MMcLGFhW5nw?feature=share)
> 
> [Stuttering Dialog Codesandbox (this is without the hidden input element)](https://codesandbox.io/p/sandbox/convoy-reproduction-environment-forked-x7j9sd?file=%2Fsrc%2Fcomponents%2Felements%2FModal.tsx&selection=%5B%7B%22endColumn%22%3A40%2C%22endLineNumber%22%3A56%2C%22startColumn%22%3A40%2C%22startLineNumber%22%3A56%7D%5D&workspace=%257B%2522activeFileId%2522%253A%2522clan5wgyn00pil1ipfe9p5ja0%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fcomponents%252Felements%252FModal.tsx%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522key%2522%253A%2522clasrdalq006o3b6iowtn6unl%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D)
> 
> **==Back to the main bug in this issue==** I tried your Codesandbox link (https://oo18kq-5173.preview.csb.app/) and the unmounting bug isn't present this time!
> 
> So seems like this is an issue with the `<Transition>` component? I'll maybe have to also do some research on the implementation in HeadlessUI.
> 
> Here's the [video demonstration](https://youtube.com/shorts/I9xu3uHNbpQ?feature=share) with your Codesandbox link that omitted the transitions.
> 
> This is a very peculiar problem, and I appreciate your help on this! Thanks again!

Let me know if you solved this!

**Comment by janhesters** [comment]
Maybe related to https://github.com/tailwindlabs/headlessui/issues/1417.

**Comment by ericwang401** [comment]
> Maybe related to #1417.

I kind of see how they're similar, but my issue occurs from focusing on inputs. Thanks for the suggestion! :)

**Comment by RobinMalfait** [comment]
@janhesters @ericwang401 they are sort of related but for different reasons. If you are using the `Dialog` component with `Transition`, then the `Transition` component knows when it is safe to mount/unmount the `Dialog`, it controls it state. The reason for that is because we want to be able to unmount with a transition but in React that's not really possible without a wrapping component that delays the "unmount" of the child.

Issue in this case is that the `Transition` component never ever finishes (instead a `transitioncancel` event happens first on Android for whatever reason), and therefore the `Dialog` never ever unmounts, which is why it looks like everything is "stuck" because we have some logic that makes the page `inert` and scroll locks as well, this is all cleaned up when the `Dialog` state changes from `open` to `closed`. But since the `Transition` component never finished, that also never happened.

Long story short, this should be fixed by #2071, and will be available in the next release.

You can already try it using `npm install @headlessui/react@insiders`.
